default.nix

```diff
@@ -220,13 +237,23 @@ let
   '';
 in emacs.overrideAttrs (esuper:
   let
+    # `--init-directory` is supported by Emacs 29+ only
+    initDirArgs = lib.optionalString isEmacs29 ''
+      if [[ $(basename $1) == emacs ]] || [[ $(basename $1) == emacs-* ]]; then
```

Comment:
Add the flag only for `emacs` and `emacs-<version>` (e.g.: `emacs-29.0.50`) binaries. There is also `emacsclient`, however this one doesn't accept this flag (and it doesn't even make sense anyway).